### PR TITLE
Storyshots: First-class CSF support

### PR DESCRIPTION
**Pull request by Hypnosphi**
Issue: #7697 

## What I did
– Migrated sample stories in `storyshots-core` to CSF
– ~~Used `displayName` in generated test names~~ reverted in favor of https://github.com/storybookjs/storybook/pull/7878#issuecomment-532397143
– Added detection of `fileName` from `req`. It won't work with `babel-plugin-require-context-hook` until https://github.com/smrq/babel-plugin-require-context-hook/pull/9 gets merged and released. Same for `require-context.macro` https://github.com/storybookjs/require-context.macro/pull/13
– Added `@storybook/addon-storyshots/injectFileName` transform that adds `exports.default.patameters.fileName`. Added a warning suggesting to include this transform when `getSnapshotFileName` is called but `fileName` is falsy.

### CODE REVIEW

File: addons/storyshots/storyshots-core/src/api/snapshotsTestsTemplate.js [hunk 1]
```diff
@@ -2,11 +2,12 @@ import { describe, it } from 'global';
 import { addSerializer } from 'jest-specific-snapshot';
 
 function snapshotTest({ item, asyncJest, framework, testMethod, testMethodParams }) {
-  const { name } = item;
+  const { parameters, name } = item;
+  const displayName = (parameters && parameters.displayName) || name;
```
Comment (shilman) [inline addons/storyshots/storyshots-core/src/api/snapshotsTestsTemplate.js]: Is there a test case that doesn't have a parameters.displayName attached? I didn't see one.

File: addons/storyshots/storyshots-core/stories/required_with_context/Button.stories.js [hunk 2]
```diff
@@ -1,18 +1,30 @@
...
-        😀 😎 👍 💯
-      </span>
-    </Button>
-  ));
+  },
+};
+
+export const withText = () => <Button onClick={action('clicked')}>Hello Button</Button>;
+
+withText.story = {
```
Comment (shilman) [inline addons/storyshots/storyshots-core/stories/required_with_context/Button.stories.js]: Remove this?

File: addons/storyshots/storyshots-core/src/api/snapshotsTestsTemplate.js (hunk 1 above)
Comment (shilman) [reply addons/storyshots/storyshots-core/src/api/snapshotsTestsTemplate.js]: I think this needs:

```
import { storyNameFromExport } from '@storybook/router/utils';

const displayName = (parameters && parameters.displayName) || storyNameFromExport(name);
```

If we want it to match what the user sees in Storybook. I'm guessing it'll work fine without.

**Review by shilman — CHANGES_REQUESTED** [review]
This looks great @Hypnosphi 💯 

I had a few minor changes inline -- should be an easy fix.

Also, I'm wondering if you've seen this PR that's vaguely related: https://github.com/storybookjs/storybook/pull/7330

I haven't had a chance to address it yet, but I'm wondering if it influences this PR at all? (It might not, just asking.)

File: addons/storyshots/storyshots-core/src/api/snapshotsTestsTemplate.js (hunk 1 above)
Comment (Hypnosphi) [reply addons/storyshots/storyshots-core/src/api/snapshotsTestsTemplate.js]: I just copypasted it from https://github.com/storybookjs/storybook/blob/next/lib/ui/src/components/sidebar/SidebarItem.tsx#L139

Comment (shilman) [reply addons/storyshots/storyshots-core/src/api/snapshotsTestsTemplate.js]: Oh you might be right then. My bad!

Comment (shilman) [reply addons/storyshots/storyshots-core/src/api/snapshotsTestsTemplate.js]: If you're getting it out of the story store then you don't need to do what I said. 👍

File: addons/storyshots/storyshots-core/stories/required_with_context/Button.stories.js (hunk 2 above)
Comment (shilman) [reply addons/storyshots/storyshots-core/stories/required_with_context/Button.stories.js]: To test out the "no displayName" case

File: addons/storyshots/storyshots-core/src/api/snapshotsTestsTemplate.js (hunk 1 above)
Comment (Hypnosphi) [reply addons/storyshots/storyshots-core/src/api/snapshotsTestsTemplate.js]: It works as is, see https://github.com/storybookjs/storybook/pull/8000/commits/88863c112749cd6c80932248d381c7d30c9cbe38

**Comment by shilman** [comment]
@Hypnosphi are there tests to make sure it still works well with the storiesOf files?

**Comment by Hypnosphi** [comment]
Ok, I'll revert some of the stories

**Comment by kylemh** [comment]
Merged https://github.com/storybookjs/require-context.macro/pull/13 and released as v1.2.0

**Comment by shilman** [comment]
@kylemh anything required on the storybook side, or is that just an FYI? AFAIK it's not a storybook dep

**Comment by kylemh** [comment]
@shilman nothing needed more from us. Just giving notice!

Note Hypno's 3rd bullet in OP.

**Comment by Hypnosphi** [comment]
@shilman it's one of the solutions to `require.context` problem we [suggest](https://github.com/storybookjs/storybook/tree/next/addons/storyshots/storyshots-core#option-2-macro) in our docs

**Comment by shilman** [comment]
@kylemh @Hypnosphi thanks for the clarification. i know about the package -- just missed the ref in your original PR @Hypnosphi . makes sense! 👍

**Comment by Hypnosphi** [comment]
> Used displayName in generated test names

As 5.2.0 is now stable, I have to make it an opt-in to avoid a breaking change

**Comment by sergej-s** [comment]
Guys, could you help me understand if this PR addresses the next issues Im facing:
1. Storyshot gets the name from `export` not from `name` even if the name is provided. I.e.
```
styledOutside.story = {
  name: 'Styled outside',
};
```
I get the storyshot `Styled Outside 1` instead of `Styled outside 1` which breaks the tests.
Does `Used displayName in generated test names` resolve this issue?

2. I used `require-context.macro` (Option 2) as described in docs but unfortunately it didn't work with CSF format. I had to resort to `babel-plugin-require-context-hook` which works good.
Does `Added detection of fileName from req` fix the issue with `require-context.macro`?

**Comment by Hypnosphi** [comment]
1. Yes, initially I've added a fix for that, but right now it would be a breaking change, so it has to be an opt-in in 5.3.0. Maybe not even on storyshots level, see https://github.com/storybookjs/storybook/pull/7878#issuecomment-532397143

2. Which errors did you get? The fix you mention only affects `multiSnapshotWithOptions`, and it doesn't work with `babel-plugin-require-context-hook` yet. So probably you had something different

**Comment by sergej-s** [comment]
1. Thank you for clarification!
2. The issue is connected to [8122](https://github.com/storybookjs/storybook/issues/8122). I wasn't able to load stories and make tests working - got the error 
image

but I managed to fix it. All this from addon-storyshots documentation:
```javascript
import { configure } from '@storybook/react';
import requireContext from 'require-context.macro';

const req = requireContext('../src/components', true, /\.stories\.js$/);
 
function loadStories() {
  req.keys().forEach((filename) => req(filename));
}

configure(loadStories, module);
```

should be replaced with this

```javascript
import { configure } from '@storybook/react';
import requireContext from 'require-context.macro';

configure(requireContext('../src/components', true, /\.stories\.js$/), module);
```

There seems to be a mismatch in documentation of storybook and addon-storyshots.

**Comment by Hypnosphi** [comment]
Looks like the documentation @sergej-s reffering to uses instructions for old story format.

@shilman should we update it?

**Comment by shilman** [comment]
@Hypnosphi @sergej-s Yes, that would be great, or I'd be happy to do it.

File: addons/storyshots/storyshots-core/src/Stories2SnapsConverter.js [hunk 3]
```diff
@@ -24,9 +24,18 @@ class DefaultStories2SnapsConverter {
   }
 
   getSnapshotFileName(context) {
-    const { fileName } = context;
+    const { fileName, kind } = context;
 
     if (!fileName) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        `Storybook was unable to detect filename for stories of kind "${kind}".
```
Comment (ndelangen) [inline addons/storyshots/storyshots-core/src/Stories2SnapsConverter.js]: should use `ts-dedent` here?

**Review by ndelangen — APPROVED** [review]
the indenting of the console.warn could be better if you use `ts-dedent` but other then that, this looks 💯 to me!

File: addons/storyshots/storyshots-core/src/Stories2SnapsConverter.js (hunk 3 above)
Comment (Hypnosphi) [reply addons/storyshots/storyshots-core/src/Stories2SnapsConverter.js]: But this is a `.js` not a `.ts` file. Is it OK?

**Comment by Hypnosphi** [comment]
@shilman can you please re-review?

**Review by shilman — APPROVED** [review]
Really great work @Hypnosphi. Thanks so much for taking care of this 🙏

**Comment by shilman** [comment]
@Hypnosphi can you take a look at the failing story in Chromatic?

File: addons/storyshots/storyshots-core/README.md [hunk 4]
```diff
@@ -588,6 +588,16 @@ Like the default, but allows you to specify a set of options for the renderer, j
 ### `multiSnapshotWithOptions(options)`
 
 Like `snapshotWithOptions`, but generate a separate snapshot file for each stories file rather than a single monolithic file (as is the convention in Jest). This makes it dramatically easier to review changes. If you'd like the benefit of separate snapshot files, but don't have custom options to pass, simply pass an empty object.
+If you use [Component Story Format](https://storybook.js.org/docs/formats/component-story-format/), you may also need to add an additional Jest transform to automate detecting story file names:
+```js
+// jest.config.js
+module.exports = {
+  transform: {
+    '^.+\\.stories\\.jsx?$': '@storybook/addon-storyshots/injectFileName',
```
Comment (JohnAlbin) [inline addons/storyshots/storyshots-core/README.md]: This documentation change differs than what is displayed in the warning below. https://github.com/storybookjs/storybook/pull/8000/files#diff-c4bcb815683c4c92762d0a868383a06cR38

Which one is the correct advice?

**Pull request merged**